internal/commands/root.go

```diff
@@ -10,15 +12,40 @@ import (
 	"github.com/spf13/pflag"
 )
 
+var (
+	completion string
+)
+
 // NewRootCmd returns the base root command.
 func NewRootCmd(use string, dockerCli command.Cli) *cobra.Command {
 	cmd := &cobra.Command{
 		Short:       "Docker Application",
 		Long:        `A tool to build and manage Docker Applications.`,
 		Use:         use,
 		Annotations: map[string]string{"experimentalCLI": "true"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			switch completion {
+			case "bash":
+				return cmd.GenBashCompletion(dockerCli.Out())
+			case "zsh":
+				return cmd.GenZshCompletion(dockerCli.Out())
+			case "":
+				// Actually unset
+				return nil
+			default:
+				return fmt.Errorf("%q is not a supported shell", completion)
+			}
+		},
 	}
 	addCommands(cmd, dockerCli)
+
+	cmd.Flags().StringVar(&completion, "completion", "", "Generates completion scripts for the specified shell (bash or zsh)")
+	err := cmd.Flags().MarkHidden("completion")
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "Failed to register command line options: %v", err.Error())
```

Comment:
Can't we just return a command and an error? And let the caller print the actual message?

Reply:
We can't, as the caller is `Plugin.Run`  which deine a `func(command.Cli) *cobra.Command` and does not expect errors to happen bulding commands

Reply:
Oh right... That's a pity... Ok then 👍